core/tests/ts-integration/tests/evm.test.ts

```diff
@@ -99,7 +172,62 @@ describeEvm('EVM contract checks', () => {
         expect(incrementGas).toBeLessThan(deploymentGas);
     });
 
-    test('should propagate constructor revert', async () => {
+    testEvm('Should propagate constructor revert', async () => {
         await expect(counterFactory.deploy(true)).rejects.toThrow('requested revert');
     });
-});
+
+    testEvm('Should define contractAddress for failed deployments', async () => {
+        const currentNonce = await alice.getNonce();
+        const expectedAddress = ethers.getCreateAddress({ from: alice.address, nonce: currentNonce });
+        const deployTx = await counterFactory.getDeployTransaction(true, { gasLimit: 10_000_000 });
+        const deployTxResponse = await alice.sendTransaction(deployTx);
+        try {
+            await deployTxResponse.wait();
+        } catch (e: any) {
+            const receipt: ethers.TransactionReceipt = e.receipt;
+            expect(receipt.status).toEqual(0);
+            expect(receipt.contractAddress).toEqual(expectedAddress);
+            return;
+        }
+        expect(null).fail('Expected deploy tx to be reverted');
+    });
+
+    testEvm('Should produce expected CREATE2 addresses', async () => {
+        const { abi, bytecode } = readContract('Counter:CounterFactory');
+        const create2Factory = await (
+            await new ethers.ContractFactory(abi, bytecode, alice).deploy()
+        ).waitForDeployment();
+        const factoryAddress = await create2Factory.getAddress();
+
+        const salt = ethers.getBytes('0x0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef');
+        await expect(await create2Factory.getFunction('deploy').send(salt)).toBeAccepted();
+
+        const deployArgs = counterFactory.interface.encodeDeploy([false]);
+        const initCode = ethers.concat([counterFactory.bytecode, deployArgs]);
+        const expectedAddress = ethers.getCreate2Address(factoryAddress, salt, ethers.keccak256(initCode));
+        const counterAddress = await create2Factory.getFunction('getAddress').staticCall(salt);
+        expect(counterAddress).toEqual(expectedAddress);
+
+        const txCount = await alice.provider!!.getTransactionCount(factoryAddress);
+        // L2 strips off the deployment part of a nonce, making `getTransactionCount()` always return 0 for contracts.
+        // L1 returns 2 since the nonce is initially set to 1 on deployment, and then incremented on the CREATE2 call.
+        const expectedTxCount = providerKind === ProviderKind.L1_ETHERS ? 2 : 0;
+        expect(txCount).toEqual(expectedTxCount);
```

Comment:
@0xVolosnikov Is returning 0 from `eth_getTransactionCount` for contracts OK? We [have a TODO](https://github.com/matter-labs/zksync-era/blob/6de1014c659247d3c29d349d5771778424241ecd/core/node/api_server/src/web3/namespaces/eth.rs#L478-L481) to return the deployment nonce for contracts, which I'd guess would be reasonably easy to implement by checking whether the queried address has an associated bytecode.